Add readonly indicator

andrewfulton9 opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References
This is a update of #12826, which fixes #7980.

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

I've reworked the status indicator to match wireframes in #7980. I also added a notification for any opened read-only files, though they may not be as loud as wanted.

<!-- Describe the code changes and how they address the issue. -->

## User-facing changes

Adds a status indicator next to debug button/kernal status. Also adds a notification if a read-only permissioned file is opened.

![image](https://github.com/jupyterlab/jupyterlab/assets/13774419/e4a9e462-698a-4f80-8484-3f7421660b54)
Above is a read-only notebook with status indicator and the notification pop-up.

![image](https://github.com/jupyterlab/jupyterlab/assets/13774419/a8cd00fe-a779-41a4-b41f-92b9ef8999d3)
Here is the same in dark mode.

![image](https://github.com/jupyterlab/jupyterlab/assets/13774419/eb776314-c13d-4507-bbe4-ce096365e0c0)
And this is what the read-only file indicator on a file looks like.

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->
None that I am aware of.

krassowski reviewed (changes requested):
Thank you for picking this up! When testing on Binder I noticed that this logic does not seem to work well in RTC mode as it shows that notebook is readonly even though it is perfectly writable:

![Screenshot from 2023-05-27 15-12-30](https://github.com/jupyterlab/jupyterlab/assets/5832902/a73dbcab-05bd-4195-9d98-de72354106ad)

On the other hand, when it is actually needed, this is in readonly source preview that can be opened from debugger, the indicator does not show up:

![file-not-marked-as-readonly](https://github.com/jupyterlab/jupyterlab/assets/5832902/5a883778-06b6-422b-9ea7-81ba0faa34bf)

andrewfulton9 commented:
@krassowski, can you take another look at this when you get a chance? I changed up where the read-only indicator is defined and implemented to be all in the docregistry extension. This allowed me to keep all the logic regarding file vs notebook handling, and RTC all in the same place instead of duplicating code. To do this I had to switch it to adding the indicator to the toolbar via `Toolbar.addItem` or `Toolbar.insertBefore`after the fact rather than defining it the toolbar schema for the file and notebook editors each respectively. For now, I am just by-passing the indicator when RTC is being used, since the autosaving functionality was causing the files to all appear as read-only. Currently in RTC, the autosave fails semi-silently (shows up in the logs, but not in the jupyterlab UI) when a file is read-only which likely should be its own issue.

andrewfulton9 commented:
Also, should `yarn.lock` be updated in a commit on the PR, or is that done as part of CI?

krassowski commented on packages/ui-components/style/toolbar.css:
```diff
@@ -101,6 +102,11 @@ button.jp-ToolbarButtonComponent .jp-ToolbarButtonComponent-label {
 
+.jp-ToolbarLabelComponent {
+  background-color: var(--jp-brand-color1);
+  color: white;
```
Does it look good in both light and dark mode? Can you add screenshots to the top-level comment?

I have not tried it yet as the Binder is flaky.

andrewfulton9 replied:
No, I'll give that a try and post the screen shots

andrewfulton9 replied:
I updated this to `var(--jp-ui-inverse-font-color1)`. Let me know what you think. I added in screenshots to the top-level-comment that you can see.

krassowski reviewed (commented):
This looks like it's no on the right track, noting some minor things to address.

andrewfulton9 commented on packages/docregistry/src/default.ts:
```diff
@@ -616,6 +616,25 @@ export class DocumentWidget<
     }
+    if (!this.context.model.dirty) {
+      const path = this.context.contentsModel?.path;
+      if (path && !path.startsWith('RTC:')) {
```
No, that looks a lot cleaner though. I'll give it a try.

andrewfulton9 replied:
This has been updated to use `this.context.model.collaborative`. Thanks for pointing that out!

andrewfulton9 commented on packages/notebook/src/panel.ts:
```diff
@@ -266,8 +259,6 @@ export class NotebookPanel extends DocumentWidget<Notebook, INotebookModel> {
     });
   }
 
-  translator: ITranslator;
```
This has been added back in

andrewfulton9 commented:
@krassowski, could you take another look at this when you get a chance?

krassowski commented on packages/docregistry/src/default.ts:
```diff
@@ -616,6 +616,24 @@ export class DocumentWidget<
+          if (!roi) {
+            this.toolbar.addItem('read-only-indicator', readOnlyIndicator);
+          }
+          Notification.warning(
+            this._trans.__(`%1 is a read-only document.`, this.title.label)
+          );
```
I think that it would be better UX if this notification was only shown on an attempt to save with <kbd>ctrl</kbd> + <kbd>s</kbd>. See this fragment of Isabela's comment https://github.com/jupyterlab/jupyterlab/issues/7980#issuecomment-655703618 (emphasis mine):

>With toast notifications moving forward in 689, I think this could be a good opportunity to remind users a notebook is read-only **if they try to save**.

Further, currently the notification does not show up but just lands in the notification centre (makes the bell in status bar blue and requires two clicks to dismiss). As it would be responding to user action I think it should be shown outright.

Nice to have: it could also include a button for "Save as" command.

andrewfulton9 replied:
I'm not sure how to deal with the notification on Ctrl + s without enabling the save commandID. Personally, I think having it disabled is more clear than enabling it and notifying if the user clicks. I did update to remove the notification, updated the caption that shows up when hovering over the save button and I added a dialog that pops up if the "save all" command is given.

What do you think? Is there a way to keep the save command disabled but still clickable?

krassowski replied:
Thank you. FYI https://github.com/jupyterlab/jupyterlab/pull/14664 changes `isEnabled` for toolbar commands which will make it always enabled even if it appears as disabled in the menu. I think we can proceed with implementation as-is and see if it needs adjusting from user feedback.

krassowski commented on packages/docregistry/src/components.tsx:
```diff
@@ -0,0 +1,33 @@
+      <span
+        className="jp-ToolbarLabelComponent"
+        title={trans.__(
+          `document is permissioned readonly; "save" is disabled, use "save as..." instead`
```
```suggestion
          `Document is permissioned readonly; "save" is disabled, use "save as..." instead`
```

krassowski reviewed (commented):
My only remaining concern is on the UX level with implementation of the notification. Otherwise it seems good to go.

andrewfulton9 commented on packages/ui-components/style/toolbar.css:
```diff
@@ -101,6 +102,11 @@ button.jp-ToolbarButtonComponent .jp-ToolbarButtonComponent-label {
   font-family: var(--jp-ui-font-family);
 }
 
+.jp-ToolbarLabelComponent {
```
Sounds good. I added this.

andrewfulton9 commented:
@fcollonval, would you mind giving this PR a review when you have a chance? Thank you!

krassowski commented on packages/docregistry/src/default.ts:
```diff
@@ -616,6 +617,21 @@ export class DocumentWidget<
     if (args.name === 'dirty') {
       this._handleDirtyState();
     }
+    if (!this.context.model.dirty) {
```
Can you remind my what the `dirty` check does here?

andrewfulton9 replied:
If I remember correctly if `this.context.model.dirty === true` then `this.context.contentsModel.writable` could also be true even if it was a read only file or vice-versa (I forget which). By adding this check here, I'm making sure it doesn't try to add a read only label until the context is fully initialized.

krassowski commented on packages/ui-components/style/toolbar.css:
```diff
@@ -101,6 +102,12 @@ button.jp-ToolbarButtonComponent .jp-ToolbarButtonComponent-label {
+.jp-ToolbarLabelComponent {
+  background-color: var(--jp-brand-color1);
+  color: var(--jp-ui-inverse-font-color1);
+  font-size: var(--jp-ui-font-size0);
```
One tiny addition:

```suggestion
  font-size: var(--jp-ui-font-size0);
  user-select: none;
```

To avoid inconsistent look when user accidentally highlights it:

![Screenshot from 2023-08-29 21-33-07](https://github.com/jupyterlab/jupyterlab/assets/5832902/f91a2d69-6ee6-49de-96d7-0d8902a08298)

krassowski reviewed (approved):
Thank you @andrewfulton9.

I opened https://github.com/jupyterlab/jupyterlab/issues/15040 to track follow-up improvements to the logic. I think this PR can be merged as it adds the widget scaffold which can be re-used in the remaining scenarios.
